feat: Adds structure for gRPC client metrics instrumentation

DNVindhya opened the pull request:
This PR is one of the initial efforts to support gRFC [A66: OpenTelemetry Metrics](https://github.com/grpc/proposal/blob/master/A66-otel-stats.md) in gRPC-Spring. Even though the gRFC mentions OpenTelmetry, to keep it in sync with existing metrics, this PR as well as future ones will use Micrometer for instrumentation.

This PR adds the structure to collect gRPC metrics for [ClientInterceptor](https://grpc.github.io/grpc-java/javadoc/io/grpc/ClientInterceptor.html). To start with I have only added instrumentation for a single metric i.e `grpc.client.attempt.started`.

fengli79 commented on grpc-client-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/client/metrics/MetricsClientInterceptor.java:
```diff
@@ -0,0 +1,74 @@
+    public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
+            MethodDescriptor<ReqT, RespT> method, CallOptions callOptions, Channel next) {
+
+        final MetricsClientStreamTracers.CallAttemptsTracerFactory tracerFactory =
```
Why create a tracer factory for each intercepted call? I'm under the impression that the factory can be reused to create a dedicated tracer for each call.

DNVindhya replied:
Along with creating tracer factory per call, we also record rpc attempt started metric. So it cannot be reused across calls.

fengli79 replied:
SG, worth to clarify that this is a per call stream tracer factory, which creates stream tracers for each attempt under the same call, each call needs a dedicated factory, as they share the same method descriptor.

DNVindhya replied:
Added comment to clarify the explicit usage of factory in every call.

fengli79 commented on grpc-client-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/client/metrics/MetricsClientInterceptor.java:
```diff
@@ -0,0 +1,74 @@
+        ClientCall<ReqT, RespT> call =
+                next.newCall(method, callOptions.withStreamTracerFactory(tracerFactory));
+
+        return new SimpleForwardingClientCall<ReqT, RespT>(call) {
```
Please leave a TODO to clarify the purpose to wrap it with a SimpleForwardingClientCall here. I'm under the impression that this is for intercept the response/error events for other metrics in the future.

DNVindhya replied:
Added TODO. You are right, it will be used to intercept metrics instrumentation on call completion event.

ST-DDT replied:
IMO the TODO comment is not very clear to me.

> // TODO(dnvindhya): Add purpose of wrapping with SimpleForwardingClientCall

Maybe change it to:

> // TODO(dnvindhya): Collect the actual response/error in the SimpleForwardingClientCall

DNVindhya replied:
Updated

fengli79 commented on grpc-client-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/client/metrics/MetricsClientStreamTracers.java:
```diff
@@ -0,0 +1,79 @@
+public class MetricsClientStreamTracers {
+
+    private static final class ClientTracer extends ClientStreamTracer {
+        final CallAttemptsTracerFactory attemptsState;
```
Please make them private as well.

DNVindhya replied:
Updated

fengli79 commented on grpc-client-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/client/metrics/MetricsClientStreamTracers.java:
```diff
@@ -0,0 +1,79 @@
+            this.fullMethodName = checkNotNull(fullMethodName, "fullMethodName");
+            this.metricsMeters = checkNotNull(metricsMeters, "metricsMeters");
+
+            // Record here in case newClientStreamTracer() would never be called.
```
Why newClientStreamTracer would never be called? Shall it be always invoked?

DNVindhya replied:
IIUC stream creation does not happen as soon as the call starts. If the call gets cancelled, even before stream is created `newClientStreamTracer()` will not be called.

fengli79 commented on grpc-client-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/client/metrics/MetricsClientStreamTracers.java:
```diff
@@ -0,0 +1,79 @@
+
+        @Override
+        public ClientStreamTracer newClientStreamTracer(StreamInfo info, Metadata metadata) {
+            if (attemptsPerCall.get() > 0) {
```
Why we need this atomic? Why not increment the attempt counter when a new stream tracer get created?

DNVindhya replied:
As soon the client call is created, it counts as an attempt. So as answered in the other [comment](https://github.com/grpc-ecosystem/grpc-spring/pull/1021/files#r1436584316) there is no certainty that `newClientStreamTracer()` will be called without fail.

ST-DDT replied:
IMO would be more clear, if this would use a boolean instead.

````java
// The first call is already counted in the constructor
if (first) {
    this.metricsMeters.getAttemptCounter()
                        .withTags(Tags.of("grpc.method", fullMethodName))
                        .increment();
} else {
  first = false;
}
````

Since this method ends afterwards, the value doesn't have to be atomic AFAICT.

DNVindhya replied:
Updated to use boolean for now. PTAL

fengli79 commented on grpc-client-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/client/metrics/MetricsMeters.java:
```diff
@@ -0,0 +1,56 @@
+/*
+ * Collection of metrics meters.
+ */
+public class MetricsMeters {
```
Can we remove this abstraction on top of MeterProvider<Counter>? Seems overweighted.

DNVindhya replied:
Added this abstraction to facilitate two things:
1. Provides ability to set individual meters which helps in selectively enabling or disabling specific metrics.
2. Acts as a data class to hold all metric metrics.
Let me know if you still think this seems redundant.

fengli79 replied:
LGTM

ST-DDT commented on grpc-client-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/client/metrics/MetricsClientInterceptor.java:
```diff
@@ -0,0 +1,74 @@
+ */
+public class MetricsClientInterceptor implements ClientInterceptor {
+
+    private final MeterRegistry registry;
```
AFAICT the registry is only used inside the constructor, so there is no need to store it in a variable.

ST-DDT commented on grpc-client-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/client/metrics/MetricsClientInstruments.java:
```diff
@@ -0,0 +1,39 @@
+    static final String CLIENT_ATTEMPT_STARTED = "grpc.client.attempt.started";
+
+    static MetricsMeters micrometerInstruments(MeterRegistry registry) {
+        MetricsMeters.Builder builder = MetricsMeters.newBuilder();
```
IMO you could just inline this:

````java
return MetricsMeters.newBuilder()
    ...
    .build();
````

DNVindhya replied:
The builder will be used to add future metrics counters and timers. So leaving it as if for now.

ST-DDT commented on grpc-client-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/client/metrics/MetricsClientStreamTracers.java:
```diff
@@ -0,0 +1,79 @@
+
+            // Record here in case newClientStreamTracer() would never be called.
+            this.metricsMeters.getAttemptCounter()
+                    .withTags((Tags.of("grpc.method", fullMethodName)))
```
AFAICT here are redundant parenthesis.

ST-DDT commented:
What is the difference between this and https://github.com/grpc-ecosystem/grpc-spring/pull/1010?

DNVindhya commented on grpc-client-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/client/metrics/MetricsClientInstruments.java:
```diff
@@ -0,0 +1,39 @@
+ */
+public class MetricsClientInstruments {
+
+    static final String CLIENT_ATTEMPT_STARTED = "grpc.client.attempt.started";
```
Made it package private so that it can accessible for tests as well. I think we can define it again in tests.
Updated access modifier and added comment.

DNVindhya commented on grpc-client-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/client/metrics/MetricsClientInstruments.java:
```diff
@@ -0,0 +1,39 @@
+
+    static final String CLIENT_ATTEMPT_STARTED = "grpc.client.attempt.started";
+
+    static MetricsMeters micrometerInstruments(MeterRegistry registry) {
```
Done

DNVindhya commented on grpc-client-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/client/metrics/MetricsClientInstruments.java:
```diff
@@ -0,0 +1,39 @@
+        MetricsMeters.Builder builder = MetricsMeters.newBuilder();
+
+        builder.setAttemptCounter(Counter.builder(CLIENT_ATTEMPT_STARTED)
+                .description("The total number of RPC attempts started, including those that have not completed.")
```
The description is from A66 [metric description](https://github.com/grpc/proposal/blob/master/A66-otel-stats.md#client-per-attempt-instruments). Updated.

DNVindhya commented on grpc-client-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/client/metrics/MetricsClientInterceptor.java:
```diff
@@ -0,0 +1,74 @@
+import io.micrometer.core.instrument.MeterRegistry;
+
+/**
+ * A gRPC client interceptor that will collect gRPC metrics.
```
Done

DNVindhya commented on grpc-client-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/client/metrics/MetricsClientInterceptor.java:
```diff
@@ -0,0 +1,74 @@
+    private final MetricsMeters metricsMeters;
+
+    /**
+     * Creates a new gRPC client interceptor that will collect metrics into the given
```
Done

DNVindhya commented on grpc-client-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/client/metrics/MetricsClientInterceptor.java:
```diff
@@ -0,0 +1,74 @@
+     *
+     * @param meterRegistry The meter registry to use.
+     */
+    public MetricsClientInterceptor(MeterRegistry meterRegistry) {
```
Done

DNVindhya commented on grpc-client-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/client/metrics/MetricsClientInterceptor.java:
```diff
@@ -0,0 +1,74 @@
+     * Creates a new gRPC client interceptor that will collect metrics into the given
+     * {@link io.micrometer.core.instrument.MeterRegistry}.
+     *
+     * @param meterRegistry The meter registry to use.
```
Done

DNVindhya reviewed (commented):
Addressed comments. PTAL

> What is the difference between this and #1010?
Accidentally rebased in the branch, and that pulled all other merged changes from master as well. So left it as draft (to be removed once this PR is reviewed) and started this PR.

ST-DDT commented on grpc-client-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/client/metrics/MetricsClientInstruments.java:
```diff
@@ -0,0 +1,46 @@
+     */
+    private static final String CLIENT_ATTEMPT_STARTED = "grpc.client.attempt.started";
+
+    static MetricsMeters instruments(MeterRegistry registry) {
```
I would probably go for `newMeters` or `newMetricsMeters` as `instruments` doesn't explain well what this method does.

DNVindhya replied:
Make sense. How about `newClientMetricsMeters`?

ST-DDT replied:
Sounds good.

DNVindhya replied:
Updated

ST-DDT commented on grpc-client-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/client/metrics/MetricsClientStreamTracers.java:
```diff
@@ -0,0 +1,79 @@
+                        .withTags((Tags.of("grpc.method", fullMethodName)))
+                        .increment();
+            }
+            if (!info.isTransparentRetry()) {
+                attemptsPerCall.incrementAndGet();
+            }
```
IMO this should either wrap around the actual increment or not be relevant at all.

DNVindhya replied:
Removed it as this is not needed for `grpc.client.attempt.started` metric.

ST-DDT commented on grpc-client-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/client/metrics/MetricsClientStreamTracers.java:
```diff
@@ -0,0 +1,79 @@
+        private final MetricsMeters metricsMeters;
+        private final AtomicLong attemptsPerCall = new AtomicLong();
+
+        CallAttemptsTracerFactory(String fullMethodName,
+                final MetricsMeters metricsMeters) {
```
```suggestion
        CallAttemptsTracerFactory(String fullMethodName, MetricsMeters metricsMeters) {
```

DNVindhya replied:
Done

ST-DDT commented on grpc-client-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/client/metrics/MetricsClientInstruments.java:
```diff
@@ -0,0 +1,46 @@
+/*
+ * The instruments used to record metrics on client.
+ */
+public class MetricsClientInstruments {
```
Please make these container classes final with a private constructor.

DNVindhya replied:
Updated

ST-DDT commented on grpc-client-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/client/metrics/MetricsClientStreamTracers.java:
```diff
@@ -0,0 +1,79 @@
+ * <p>
+ * On the client-side, a factory is created for each call, and the factory creates a stream tracer for each attempt.
+ */
+public class MetricsClientStreamTracers {
```
Same here

DNVindhya replied:
Updated

ST-DDT reviewed (approved):
Implementation wise this looks good to me.

This is missing for the inclusion in an actual release:

- the metric auto configuration
- and some tests

My concerns with this PR:

- This is only a partial implementation of the entire metrics feature
- This should be implemented in the micrometer grpc instrumentation repo (where there are already two different implementations of it)

DNVindhya commented:
Thanks for all your detailed comments. This PR was added with the intention to get the basic structure in place to add a single client metric, other metrics and unit tests will be added in subsequent PRs. Plan is to enable metric auto configuration, once all the metrics are in.

> * This should be implemented in the micrometer grpc instrumentation repo (where there are already two different implementations of it)

Since I am breaking down the metrics features (for ease of review and faster iteration) into multiple PRs, I am planning to move these to Micrometer GRPC instrumentation repo after the feature is completely checked in and ready to be enabled. 

I hope I addressed your concerns. Please let me know if there are any blockers to merge the PR.

ST-DDT commented:
> Please let me know if there are any blockers to merge the PR.

All good. LGTM.
